pkg/jobmanager/jobmanager.go

```diff
@@ -156,114 +161,98 @@ loop:
 		select {
 		// handle events from the API
 		case ev := <-a.Events:
-			ev.Context.Debugf("Handling event %+v", ev)
+			ctx.Debugf("Handling event %+v", ev)
 			// send the response, and wait for the given timeout
-			jm.handleEvent(ev)
+			jm.handleEvent(ctx, ev)
 		// check for errors or premature termination from the listener.
 		case err := <-errCh:
-			log.Infof("JobManager: API listener failed, triggering a cancellation of all jobs")
-			jm.CancelAll(ctx)
 			if err != nil {
-				return fmt.Errorf("error reported by API listener: %v", err)
-			}
-			return errors.New("API listener terminated prematurely without errors")
-		// handle signals to shut down gracefully. If the cancellation takes too
-		// long, it will be terminated.
-		case sig := <-sigs:
-			// TODO: stop processing signals inside jobmanager, this is
-			//       as responsibility of "main".
-			// We were interrupted by a signal, time to leave!
-			if sig == syscall.SIGUSR1 {
-				log.Debugf("Interrupted by signal '%s': wait for jobs and exit", sig)
-				apiCancel()
-			} else {
-				log.Debugf("Interrupted by signal '%s': pause jobs and exit", sig)
-				apiCancel()
-				jm.PauseJobs(ctx)
-			}
-			select {
-			case err := <-errCh:
-				if err != nil {
-					return fmt.Errorf("API listener terminated with error: %v", err)
-				}
-				// break the outer loop
-				break loop
-			case <-time.After(cancellationTimeout):
-				return fmt.Errorf("API listener didn't shut down within %v, exiting", cancellationTimeout)
+				ctx.Infof("JobManager: API listener failed (%v)", err)
 			}
+			break loop
+		case <-ctx.Until(xcontext.ErrPaused):
+			ctx.Infof("Paused")
+			jm.PauseAll(ctx)
+			break loop
+		case <-ctx.Done():
+			break loop
+		}
+	}
+	// Stop the API (if not already)
+	jm.StopAPI()
+	for !jm.checkIdle(ctx) {
+		select {
+		case <-ctx.Done():
+			ctx.Infof("Canceled")
+			jm.CancelAll(ctx)
+		case <-time.After(1 * time.Second):
```

Comment:
What are we waiting for here? (can you put in a comment?)

Reply:
we are waiting for jobs to finish or for pause to be upgraded to cancellation. added comments.